### DISCUSSION TITLE
Support for Stepper motor - Where to start??

### DISCUSSION THREAD

**Discussion by techboycr** [iNavFlight/inav#6921] [Q&A]
I been using iNav for quite sometime in quads and wings, but this time I want to use it with a self balancing robot. This design come from jjrobots: https://www.instagram.com/p/COa9GmLBy8a/

So I am wondering where to start adding support for this feature, pretty much I have to:

1. Grab filtered readings
2. Create timers for Stepper
3. Add the balance algorithm, pretty much rewrite the jjrobots implementation.

The way it works right now is just like the quad implementation, it sets an angle and corrects until the set point is reach.

Any help in how to start will be highly appreciated. Thanks!!

**Comment by DzikuVx** [comment]
INAV does not and probably will not support steppers directly connected:
1. Huge effort
2. one stepper driver requires 2 pins 
3. 
My suggestion is to develop a stepper driver that can be driven by a PWM signal. Think about a ESC do steppers: it accepts PWM and converts to stepper specific. More less, a "big servo". 
Such a driver can be build with arduino or any other microcontroller and generic stepper drivers like A4988

**Reply by techboycr** [reply]
@DzikuVx  you so much! And huge thanks for all the work you and every developer puts on iNav.

That was my first guess, create one driver that can be actuated like a 3D brushless driver. So i will continue that route.

Is there any documentation regarding iNav code structure? So I can better understand where to start and add support for this driver.

Thanks!!

**Reply by techboycr** [reply]
My setup for the driver is a bluepill (STM32F1) and TMC2100. Let's see how it goes...😅

**Comment by DzikuVx** [comment]
If it will read standard PWM signal, there is no need to change anything in
INAV. It will just output data

wt., 4 maj 2021, 17:42 użytkownik Abraham Pineda Zelaya <
***@***.***> napisał:

**Reply by techboycr** [reply]
Roger that!
